aip/0162.md

```diff
@@ -0,0 +1,229 @@
+---
+aip:
+  id: 162
+  state: draft
+  created: 2019-09-17
+permalink: /162
+redirect_from:
+  - /0162
+---
+
+# Resource Revisions
+
+Some APIs need to have resources with a revision history, where users can
+reason about the state of the resource over time. There are several reasons for
+this:
+
+- Users may want to be able to roll back to a previous revision, or diff
+  against a previous revision.
+- An API may create data which is derived in some way from a resource at a
+  given point in time. In these cases, it may be desirable to snapshot the
+  resource for reference later.
+
+## Guidance
+
+APIs **may** store a revision history for a resource if it is useful to users.
+
+APIs implementing resources with a revision history **must** provide a
+`revision_id` field on the resource:
+
+```proto
+message Book {
+  // The name of the book.
+  string name = 1;
+
+  // Other fields…
+
+  // The revision ID of the book.
+  // This is automatically changed whenever the contents of the book
+  // change in any way.
+  string revision_id = 99 [(google.api.field_behavior) = OUTPUT_ONLY];
+}
+```
+
+- The `revision_id` field **should** be a string and contain a short,
+  automatically-generated hash. A good rule of thumb is the last eight
+  characters of a UUID4.
+- The field **must** document when new revisions are created.
```

Comment:
Would there be an option that e.g.

CreateModel()  creates a totally new model when no name is provided to it as argument, while if a name is provided it would just create a new version of an existing model? Or would you say it would be better to have 2 methods instead, e.g. CreateModel() and CreateModelVersion() (I don't like this option too much).

Reply:
@jgeewax and @rsurowka and I discussed this in person, and we agreed:

- Definitely do not overload `CreateModel`.
- We think that due to varying use cases in APIs that we should be flexible in how APIs create revisions so long as it is clearly documented. In AutoML's case, `ReplaceModel` is likely to be reasonable, but `CreateModelRevision` may also make sense, and we are unsure which at this time.
  - Note: A more explicit approach is likely more future-proof, so we should probably use that as the sensible default.